src/fluree/db/json_ld/shacl.cljc

```diff
@@ -352,14 +406,15 @@
   or nil if none exist."
   [{:keys [schema] :as db} type-sids]
   (go-try
-    (let [shapes-cache (:shapes schema)]
+   (let [shapes-atom (:shapes schema)
+         cached-shapes @shapes-atom]
       (loop [[type-sid & r] type-sids
              shape-maps nil]
         (if type-sid
-          (let [shape-map (if (contains? (:class @shapes-cache) type-sid)
-                            (get-in @shapes-cache [:class type-sid])
+          (let [shape-map (if-let [cached (get-in cached-shapes [:class type-sid])]
```

Comment:
This won't quite achieve the efficiencies of the cache like the version of that was there which instead used `contains?`.

If the case exists there is no SHACL constraint for a class, the swap! will still insert 'nil' for the key such that we know it has already been looked up, but none existed.

The `if-let` here will not trigger with a `nil` value, and thus do a lookup over and over for the same class. `contains?` on the other hand will recognize the key exists (it has been looked up before) and then return its value, which might be `nil`.

Reply:
Ahh, good point, I can revert it. I only touched this part of the code because I was print-debugging the contents of the atom during implementation, I can just put it back how it was.